File: Example/Tests/StarIOTests.m
//
//  TestDomainDataServices.m
//  WDePOS
//
//  Created by Danko, Radoslav on 17/10/15.
//  Copyright © 2015 Wirecard. All rights reserved.
//


#import <XCTest/XCTest.h>
#import <libextobjc/EXTScope.h>
#import "SaleHelper.h"
#import "UserHelper.h"
#import "DDLog.h"
#import <WDePOS/NSString+WDExtras.h>
#import "BaseTests.h"

@interface StarIOTestsObjC : BaseTestsObcj <WDScanning, WDPrinting, WDManagerDelegate>
{
    __block BOOL printingSuccess;
    __block BOOL openingDrawerSuccess;
    __block BOOL openingDrawerDone;
    __block BOOL scanningSuccess;
    WDExtensionTypeUUID deviceModel;
}

@end

@implementation StarIOTestsObjC

- (void)setUp
{
    [super setUp];
    deviceModel = WDMPOPExtensionUUID;
    openingDrawerSuccess = NO;
    openingDrawerDone = NO;
    scanningSuccess = NO;
}

-(void)testStarMicronics
{
#if TARGET_OS_SIMULATOR
    NSLog(@"\n\t\t   %s Not runnable on simulator 📱",__PRETTY_FUNCTION__);
    return;
    
#endif
    
    //PART 1: We detect paired devices
    //--------------------------------------
    expectation = [self expectationWithDescription:@"We detect paired devices"];
    [self discoverDevices:deviceModel];
    [self waitForExpectationsWithTimeout:100 handler:nil];
    if (!selectedDevice)
    {
        XCTFail(@"No device found. Please make the hardware is switched on and paired to yout iOS device settings");
    }
    else
    {
        [sdk.scannerManager addScannerManagerDelegate:self
                                           forScanner:selectedDevice];
        [sdk.printerManager addPrinterManagerDelegate:self
                                           forPrinter:selectedDevice];
        [sdk.cashDrawerManager addManagerDelegate:self
                                        forDevice:selectedDevice];
    }
    
    //PART 2: We open the cash drawer
    //--------------------------------------
    expectation = [self expectationWithDescription:@"Open cash drawer"];
    //Open of cash drawer is trigger in delegate method connectionStatusDidChange below, once it is connected
    [self waitForExpectationsWithTimeout:100 handler:nil];
    if (!openingDrawerSuccess)
    {
        XCTFail(@"Device communication failed. Please make the hardware is switched on and paired to yout iOS device settings. If it was paired to other iOS device, please unpair it first.");
    }
    
    //PART 3: Printing receipt
    //--------------------------------------
    expectation = [self expectationWithDescription:@"Printing receipt"];
    [self printReceipt];
    [self waitForExpectationsWithTimeout:100 handler:nil];
    if (!printingSuccess)
    {
        XCTFail(@"Printing failed. Please make the hardware is switched on and paired to yout iOS device settings. If it was paired to other iOS device, please unpair to it first.");
    }
    
    if (deviceModel == WDMPOPExtensionUUID) //StarMicronics does not have barcode scanner
    {
        //PART 4: Scanning barcode
        //--------------------------------------
        expectation = [self expectationWithDescription:@"Read barcode"];
        [self waitForExpectationsWithTimeout:100 handler:nil];
    }
}

-(void)openCashDrawer
{
    void (^completion)(BOOL, NSError*) = ^(BOOL success, NSError *error)
    {
        if (!openingDrawerSuccess)
        {
            openingDrawerSuccess = success && !error;
            returnedErr = error;
            [expectation fulfill];
        }
    };
    
    [sdk.cashDrawerManager openCashDrawer:selectedDevice completion:completion];
}

-(void)printReceipt
{
    PrinterStateUpdate progress = ^(WDPrinterStateUpdate update)
    {
        NSLog(@"Print progress is: %ld", (long)update);
    };
    
    PrintCompletion completion  = ^(BOOL success, NSError *error) {
        printingSuccess = !error && success;
        returnedErr = error;
        [expectation fulfill];
    };
    
    WDPrinterConfig *printerConfig = [[WDPrinterConfig alloc] initWithPrinter:selectedDevice printJobs:@[[UIImage imageNamed:@"fakereceipt"]]];
    /*
     Note that the printer adds border to the image and that cannot be avoided. Try to not include
     borders also in the picture or the image will be scaled down too much.
     For mPOP and StarMicronics devices, printing from image results in more flexibility and not a big impact in speed of efficiency (Only Datecs is limited due
     to battery usage and paper size)
    
     If you wish to use the Default Design receipt you can obtain the SaleResponse from
     A. saleManager pay:
     B. saleManager querySales:
     Here we demonstrate the use of saleManager query: to obtain a SaleResponse a use it for printing the Default Sale Receipt
        Query parameters such as paging, order by , sort
        and specific filters such as full text search, sale creation date, sale operation, sale status
    NOTE: Star Micronics devices only accept receipts in the format UIImage
     */
    WDSalesQuery *query = [WDSalesQuery new];
    query.page = 0;
    query.pageSize = 20;
    query.orderBy = WDSaleQueryOrderByCreatedAt; // Order by Creation date
    query.orderSort = WDQuerySortDescending; // Sort Descending - latest Sale first
    query.statuses = @[@(WDSaleStateCompleted)]; // Only Sales with Completed status
    query.saleTypes = @[    @(WDSaleTypePurchase) ];// Purchases
    
    //query Sales
    [[sdk saleManager] querySales:query completion:^(NSArray<WDSaleResponse *> * arr, NSError * error)
     {
         saleResponse = [arr firstObject]; //Use the first returned sale
         if (saleResponse){
         
             //alternatively use the Default Designed receipt for Datecs
             [saleResponse receipt:YES //Is the receipt for cardholder use
                       showReturns:NO //Should the receipt contain returns ?
                            format:WDPrintFormatUIImage //The receipt format is UIImage object for StarIO or mPOP
                               dpi:WDPrintDpiStarMicronics //The dots per width is set to StarMicronics
                        completion:^(NSArray* _Nullable receipts, NSError* _Nullable error) {
                            //Receipt as per format specified HTML | PDF | UIImage | WDReceipt object
                            returnedErr = error;
                            printerConfig.printJobs = receipts;
                            [sdk.printerManager print:printerConfig
                                             progress:progress
                                           completion:completion];
                            
                        }];
         }
         else{
             [expectation fulfill];
         }
         returnedErr = error;
         
     }];
}

- (void)device:(WDTerminal*)device barcodeReceived:(NSString *)barcodeReceived symbology:(WDBarcodeSymbology)symbology
{
    NSLog(@"Barcode read with value as string: %@", barcodeReceived);
    if (!scanningSuccess) {
        scanningSuccess = YES;
        [expectation fulfill];
    }
}

-(void)device:(WDTerminal *)device paperSatusUpdated:(WDPrinterPaperStatus)status
{
    //You can warn the user about the level of paper by listening to this delegate call
    NSLog(@"The level of paper in your printer is %ld", status);
}

- (void)device:(WDTerminal*)device connectionStatusDidChange:(WDExtensionConnectionStatus)status
{
    NSLog(@"Connection status changed %ld", status);
    if (status == WDExtensionConnectionStatusConnected && !openingDrawerDone)
    {
        openingDrawerDone = YES;
        [self openCashDrawer];
    }
}

-(void)device:(WDTerminal *)device batteryLevelDidChange:(NSInteger)batteryLevel
{
    //Never called as this device does not report it
}

-(void)discoverDevices:(WDExtensionTypeUUID)extensionUUID
{
    DeviceDiscoveryCompletion completionTerminals = ^(NSArray *arr, NSError *err)
    {
        returnedErr = err;
        selectedDevice = arr.firstObject;
        [expectation fulfill];
    };
    
    [sdk.cashDrawerManager discoverDevices:extensionUUID completion:completionTerminals];
}
@end

File: Example/Tests/StarIOTests.swift
//
//  StatsTests.swift
//  DemoEPOS
//
//  Created by Fortes, Francisco on 7/25/17.
//  Copyright © 2017 Wirecard. All rights reserved.
//

import Foundation
import XCTest
import WDePOS

class StarIOTestsSwift: BaseTestsSwift, WDScanning, WDPrinting, WDManagerDelegate
{
    var userHelper : UserHelper!
    var printingSuccess : Bool?
    var openingDrawerSuccess, isConnected  : Bool?
    var scanningSuccess : Bool?
    var deviceModel : WDExtensionTypeUUID!
    
    override func setUp()
    {
        super.setUp()
        deviceModel = .WDMPOPExtensionUUID // Use .WDStarMicronicsExtensionUUID for printer
        self.openingDrawerSuccess = false
        self.isConnected = false
        self.scanningSuccess = false
        self.scanningSuccess = false
    }

    func testStarMicronics()
    {
        #if arch(i386) || arch(x86_64)
            let file:NSString = (#file as NSString).lastPathComponent as NSString
            NSLog("\n\t\t    [%@ %@] Not runnable on simulator 📱",file.deletingPathExtension , #function);
            return
            
        #else
        
        //PART 1: We detect paired devices
        //--------------------------------------
        expectation = self.expectation(description: "We detect paired devices")
        self.discoverDevices(deviceModel)
        self.waitForExpectations(timeout: 100, handler: nil)
        if (selectedDevice == nil)
        {
            XCTFail("No device found. Please make the hardware is switched on and paired to yout iOS device settings");
        }
        else
        {
            sdk.scannerManager.add(self, forDevice: selectedDevice!)
            sdk.printerManager.add(self, forDevice: selectedDevice!)
            sdk.cashDrawerManager.add(self, forDevice: selectedDevice!)
        }
        
        //PART 2: We open the cash drawer
        //--------------------------------------
        expectation = self.expectation(description: "Open cash drawer")
        self.waitForExpectations(timeout: 100, handler: nil)
        if (isConnected == false)
        {
            XCTFail("Device communication failed. Please make the hardware is switched on and paired to yout iOS device settings. If it was paired to other iOS device, please unpair it first.");
        }
        
        expectation = self.expectation(description: "Open cash drawer")
        self.openCashDrawer() //It will retry if failed
        self.waitForExpectations(timeout: 100, handler: nil)
        
        
        //PART 4: Getting a sale
        //--------------------------------------
        expectation = self.expectation(description: "Getting a sale")
        self.gettingSale()
        self.waitForExpectations(timeout: 100, handler: nil)
        if (saleResponse == nil)
        {
            XCTFail("No sale could be reached. Please run cash unit test first, for example.");
        }
        
        //PART 3: Printing receipt
        //--------------------------------------
        expectation = self.expectation(description: "Printing receipt")
        self.printReceipt()
        self.waitForExpectations(timeout: 100, handler: nil)
        if (printingSuccess == false)
        {
            XCTFail("Printing failed. Please make the hardware is switched on and paired to yout iOS device settings. If it was paired to other iOS device, please unpair to it first.");
        }
        
        if (deviceModel == .WDMPOPExtensionUUID) //StarMicronics CB2002 does not have barcode scanner, so disable this if you are not using mPOP one
        {
            //PART 4: Scanning barcode
            //--------------------------------------
            expectation = self.expectation(description: "Read barcode")
            self.waitForExpectations(timeout: 100, handler: nil)
        }
        
        #endif
    }
    
    func openCashDrawer()
    {
        if let selectedDevice = self.selectedDevice {
            sdk.cashDrawerManager.openCashDrawer(selectedDevice,
                                                 completion: {[weak self](success : Bool?, error : Error?) in
                                                    if error != nil || success == false
                                                    {
                                                        //We retry if failed. In your application, handle the error properly and trigger the open drawing from UI, ie. from a button. On unit tests, when print, open and scanning is all triggered and monitored on a unit tests, the StarIO device is known to fail sometimes.
                                                        self?.openCashDrawer()
                                                        return
                                                    }
                                                    self?.openingDrawerSuccess = success
                                                    self?.returnedErr = error
                                                    self?.expectation.fulfill()
            })
        }
    }
    
    func gettingSale()
    {
        let statuses = [NSNumber].init(arrayLiteral: NSNumber.init(value: WDSaleState.completed.rawValue))
        let types = [NSNumber].init(arrayLiteral:NSNumber.init(value: WDSaleType.purchase.rawValue))
        let paymentMethods = [NSNumber].init(arrayLiteral:NSNumber.init(value: WDPaymentMethod.cash.rawValue))

        let query : WDSalesQuery = WDSalesQuery.init(page: 0,
                                                                 pageSize: 20,
                                                                 orderBy: .createdAt,
                                                                 order: .descending,
                                                                 statuses: statuses,
                                                                 saleTypes: types,
                                                                 paymentMethods: paymentMethods)

        sdk.saleManager.querySales(query, completion: {[weak self](arr : [WDSaleResponse]?, error : Error?) in
            self?.returnedErr = error
            self?.saleResponse = arr?.first //Use the first returned sale for this example
            self?.expectation.fulfill()
        })
    }
    
    func printReceipt()
    {
        guard let receiptImage = UIImage.init(named: "fakereceipt") else
        {
            XCTFail("Something went really wrong - printReceipt")
            self.expectation.fulfill()
            return
        }
        
        let printProgress : PrinterStateUpdate = {(update : WDPrinterStateUpdate) in
            print("Print progress is \(update.rawValue)")
        }
        
        let printCompletion : PrintCompletion = {[weak self](success : Bool?, error : Error?) in
            self?.printingSuccess = success
            self?.returnedErr = error
            self?.expectation.fulfill()
        }
        
        guard let selectedDevice = self.selectedDevice else {
            XCTFail("No selected device to print - printReceipt")
            self.expectation.fulfill()
            return
        }

        let printerConfig : WDPrinterConfig = WDPrinterConfig.init(printer: selectedDevice,
                                                                               printJobs: [UIImage].init(arrayLiteral: receiptImage))
        /*Note that the printer adds borders to the image and it cannot be avoided. Try  not to include
        borders in the picture -- otherwise the image will be scaled down.
        If you wish to use the Default Design receipt you can obtain the SaleResponse from
            a) a saleManager pay:
            b) a saleManager querySales:
        Here we demonstrate the use of saleManager/saleResponse query.
        NOTE: Star Micronics devices only accept receipts in the format UIImage
        */

        self.saleResponse?.receipt(true,
                              showReturns: false,
                              format: .uiImage,
                              dpi: .starMicronics,
                              completion: {[weak self](receipts : [Any]?, error : Error?) in
            self?.returnedErr = error
            //Receipt as per format specified HTML | PDF | UIImage | WDReceipt object
            if let receipts = receipts
            {
                printerConfig.printJobs = receipts
                self?.sdk.printerManager.print(printerConfig,
                                               progress: printProgress,
                                               completion: printCompletion)
            }
            else
            {
                self?.expectation.fulfill()
            }
        })
    }
    
    func device(_ device: WDTerminal, connectionStatusDidChange status:WDExtensionConnectionStatus)
    {
        print("Connection status changed \(status)")
        if (status == .connected && self.isConnected == false)
        {
            self.isConnected = true
            self.expectation.fulfill()
        }
    }
    
    func device(_ device: WDTerminal, batteryLevelDidChange batteryLevel: Int)
    {
        //StarIO devices do not inform about battery levels
    }
    
    func device(_ device: WDTerminal, paperStatusUpdated status: WDPrinterPaperStatus)
    {
        print("The level of paper in your printer is \(status)")
    }
    
    func removeSpecialCharsFromString(text: String) -> String
    {
        let okayChars : Set<Character> =
            Set("abcdefghijklmnopqrstuvwxyz ABCDEFGHIJKLKMNOPQRSTUVWXYZ1234567890+-*=(),.:!_")
        return String(text.filter {okayChars.contains($0) })
    }
    
    func device(_ device: WDTerminal, barcodeReceived: String, symbology: WDBarcodeSymbology) //Data received through barcode scanner
    {
        print("Barcode read with value as string: \(barcodeReceived)")
        
        if self.scanningSuccess == false {
            self.scanningSuccess = true
            self.expectation.fulfill()
        }
    }

    func discoverDevices(_ extensionUUID: WDExtensionTypeUUID)
    {
        sdk.cashDrawerManager.discoverDevices(extensionUUID,
                                           completion: {[weak self](arr : [WDTerminal]?, error : Error?) in
                                            self?.selectedDevice = arr?.first
                                            self?.returnedErr = error
                                            self?.expectation.fulfill()
        })
    }

    
}



